pkg/plugins/runtime/k8s/controllers/workload_controller.go

```diff
@@ -0,0 +1,213 @@
+	log := r.Log.WithValues("workload", workloadName, "mesh", meshName, "namespace", namespace)
+
+	workload := &workload_k8s.Workload{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      workloadName,
+			Namespace: namespace,
+		},
+	}
+
+	result, err := kube_controllerutil.CreateOrUpdate(ctx, r.Client, workload, func() error {
+		// Set mesh label
+		if workload.Labels == nil {
+			workload.Labels = map[string]string{}
+		}
+		workload.Labels[mesh_proto.MeshTag] = meshName
+		workload.Labels[mesh_proto.ManagedByLabel] = "k8s-controller"
+
+		// Ensure spec is initialized (empty struct is fine)
+		if workload.Spec == nil {
+			workload.Spec = &workload_api.Workload{}
+		}
+
+		// Don't set status - leave it nil
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to create/update Workload %s in namespace %s", workloadName, namespace)
+	}
+
+	switch result {
+	case kube_controllerutil.OperationResultCreated:
+		log.Info("workload created")
+	case kube_controllerutil.OperationResultUpdated:
+		log.V(1).Info("workload updated")
+	}
+
+	return nil
+}
+
+func (r *WorkloadReconciler) cleanupOrphanedWorkloads(ctx context.Context, namespace string) error {
+	log := r.Log.WithValues("namespace", namespace)
+
+	// List all Workloads in the namespace
+	workloads := &workload_k8s.WorkloadList{}
+	if err := r.List(ctx, workloads, kube_client.InNamespace(namespace)); err != nil {
+		return errors.Wrap(err, "unable to list Workloads")
+	}
+
+	// List all Dataplanes in the namespace
+	dataplanes := &mesh_k8s.DataplaneList{}
+	if err := r.List(ctx, dataplanes, kube_client.InNamespace(namespace)); err != nil {
+		return errors.Wrap(err, "unable to list Dataplanes")
+	}
+
+	// Build a map of workload names that are still referenced
+	referencedWorkloads := make(map[string]map[string]bool) // mesh -> workload name -> true
+	for _, dp := range dataplanes.Items {
+		if workloadName, ok := dp.GetAnnotations()[metadata.KumaWorkload]; ok && workloadName != "" {
+			if referencedWorkloads[dp.Mesh] == nil {
+				referencedWorkloads[dp.Mesh] = make(map[string]bool)
+			}
+			referencedWorkloads[dp.Mesh][workloadName] = true
+		}
+	}
+
+	// Delete Workloads that are no longer referenced
+	for _, workload := range workloads.Items {
+		workloadName := workload.Name
+		meshName := workload.GetMesh()
+
+		if referencedWorkloads[meshName] == nil || !referencedWorkloads[meshName][workloadName] {
+			log.Info("deleting orphaned workload", "workload", workloadName, "mesh", meshName)
+			if err := r.Delete(ctx, &workload); err != nil && !kube_apierrs.IsNotFound(err) {
+				return errors.Wrapf(err, "failed to delete orphaned Workload %s", workloadName)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *WorkloadReconciler) SetupWithManager(mgr kube_ctrl.Manager) error {
+	return kube_ctrl.NewControllerManagedBy(mgr).
+		Named("kuma-workload-controller").
+		For(&mesh_k8s.Dataplane{}).
```

Comment:
I think you need to create controller for `&workload_k8s.Workload` instead of `&mesh_k8s.Dataplane{}` and watch Dataplanes. In that case `Reconcile` will be invoked with name/namespace of the workload, so you'll be able to: 
* list all DPPs for the workload (and that's how potentially you implement status later)
* delete workload when there are no DPPs for the workload (instead of constantly calling `cleanupOrphanedWorkloads` which is a rather expensive method

Reply:
Good idea i didn't thought about it earlier. Updated